File: svn/0.prealpha/software/MatLab/modelfree/locglmfit.m
function [pfit,etafit,H] = locglmfit(xfit,r,m,x,h,link,guessing,lapsing,...
    K,p,ker,maxiter,tol)
%
% Local polynomial estimator for the psychometric function (PF) and eta
% function (PF transformed by link) for binomial data; also returns the hat
% matrix.  
%
% Actual calculations are done in LOCGLMFIT_PRIVATE or
% LOCGLMFIT_SPARSE_PRIVATE depending on the size of the data set. Here the
% data are split into several parts to speed up the calculations.  
%
%
%INPUT
%
% xfit - points in which to calculate the estimate
% r - number of successes in points x
% m - number of trials in points x 
% x - stimulus levels 
% h - bandwidth(s)
%
% OPTIONAL INPUT
%
% link - name of the link function to be used; default is 'logit'
% guessing - guessing rate; default is 0
% lapsing - lapsing rate; default is 0
% K - power parameter for Weibull and reverse Weibull link; default is
% 2
% p - degree of the polynomial; default p = 1
% ker - kernel function for weights; default 'normpdf'
% maxiter - maximum number of iterations in Fisher scoring; default is 200
% tol - tolerance level at which to stop Fisher scoring; default is 1e-6
%
% OUTPUT
%
% pfit - value of the local polynomial estimate in points xfit
% etafit - estimate of eta (link of pfit)
% H - hat matrix (OPTIONAL)

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%% PROGRAM

%%%% CHECK INPUT PARAMETERS
% First 5 arguments are mandatory
if (nargin<5)
    error('Check input. First 5 arguments are mandatory');
end

%%%%
%%%% DEFAULTS
if (nargin<6)
    link = 'logit';
    disp('default link function is ''logit''');
end

if (nargin<7)
    guessing = 0;
    disp('default guessing rate is zero');
end

if (nargin<8)
    lapsing = 0;
    disp('default lapsing rate is zero');
end

if (nargin<9)
    K = 2;
    if strcmp(link, 'weibull')
        disp('default exponent for Weibull link function is 2');
    elseif strcmp(link, 'revweibull')
        disp('default exponent for reverse Weibull link function is 2');
    end
end

if (nargin<10)
    p = 1;
    disp('degree of the polynomial to be fitted on the linear scale is 1');
end

if (nargin<11)
    ker = 'normpdf';
    disp('default kernel is ''normpdf''');
end

if (nargin<12)
    maxiter=200;
    disp('default maximum number of iterations is 200');
end

if (nargin<13)
    tol = 1e-6;
    disp('default tolerance is 1e-6');
end

%%%% CHECK ROBUSTNESS OF INPUT PARAMETERS
[ rxfit, cxfit ] = size( xfit );
if cxfit ~= 1
    error('vector of x points should be column vector');
end
clear data;
data(1).content = x;
data(2).content = r;
data(3).content = m;
checkinput( 'psychometricdata2', data );
clear data
[ rh, ch ] = size( h );
if ch ~= 1
    error('vector of bandwidths should be column vector');
end
if ~( length( h ) == 1 || rh == rxfit )
    error( 'bandwidth h must be either a scalar or a vector with the same number of elements as x' );
end
checkinput( 'linkfunction', link );
if length( guessing ) > 1
    error( 'guessing rate must be scalar' );
end
if length( lapsing ) > 1
    error( 'lapsing rate must be scalar' );
end
checkinput( 'guessingandlapsing', [ guessing 1 - lapsing ] );
checkinput( 'exponentk', K );
checkinput( 'degreepolynomial', p );
checkinput( 'kernel', ker );
checkinput( 'maxiter', maxiter );
checkinput( 'tolerance', tol );

%%%%%
%%%%% INITIAL VALUES
split = 20;

Lxfit = length(xfit);
Lx = length(x);

if (Lx > 15),
    % big data
    fun_estim = 'locglmfit_sparse_private';
else
    % small data
    fun_estim = 'locglmfit_private';
end

%%%% SPLIT AND EVALUATION
if length( h ) == 1
    % with Hat matrix
    if nargout == 3,

        if (Lxfit<=split),

            % small x
            [pfit,etafit,H] = feval(fun_estim,xfit,r,m,h,x,link,guessing,lapsing,...
                K,p,ker,maxiter,tol);
        else
            % large x
            % number of parts into which the fitting is divided
            fLx = floor(Lxfit/split);

            % initialise output
            pfit = [];
            etafit = [];
            H = [];   

  
                for i = 0:(fLx-1),
                    % part of the fit
                    [pfit1,etafit1,H1] = feval(fun_estim,xfit(i*split + (1:split)),r,m,...
                        h,x,link,guessing,lapsing,K,p,ker,maxiter,tol);
                    % put the fits together
                    pfit = [pfit;pfit1];
                    etafit = [etafit;etafit1];
                    H = [H;H1];   
                end
                %final part of the fit
                if ((split*fLx)<Lxfit),
                    [pfit1,etafit1,H1] = feval(fun_estim,xfit((1+split*fLx):Lxfit),r,m,...
                        h,x,link,guessing,lapsing,K,p,ker,maxiter,tol);

                    % put the fits together
                    pfit = [pfit;pfit1];
                    etafit = [etafit;etafit1];
                    H = [H;H1];   
                end

 
        end

    else % no Hat matrix

        if (Lxfit<=split),

            % small x
            [pfit,etafit] = feval(fun_estim,xfit,r,m,h,x,link,guessing,lapsing,...
                K,p,ker,maxiter,tol);
        else
            % large x
            % number of parts into which the fitting is divided
            fLx = floor(Lxfit/split);

            % initialise output
            pfit = [];
            etafit = [];

            for i = 0:(fLx-1),
                % part of the fit
                [pfit1,etafit1] = feval(fun_estim,xfit(i*split + (1:split)),r,m,h,x,...
                    link,guessing,lapsing,K,p,ker,maxiter,tol);
                % put the fits together
                pfit = [pfit;pfit1];
                etafit = [etafit;etafit1];
            end
            %final part of the fit
            if ((split*fLx)<Lxfit),
                [pfit1,etafit1] = feval(fun_estim,xfit((1+split*fLx):Lxfit),r,m,h,x,...
                    link,guessing,lapsing,K,p,ker,maxiter,tol);

                % put the fits together
                pfit = [pfit;pfit1];
                etafit = [etafit;etafit1];
            end

        end
    
    end % if nargout == 3
else % if length( h ) == 1
    % with Hat matrix
    if nargout == 3,

        if (Lxfit<=split),

            % small x
            [pfit,etafit,H] = feval(fun_estim,xfit,r,m,h,x,link,guessing,lapsing,...
                K,p,ker,maxiter,tol);
        else
            % large x
            % number of parts into which the fitting is divided
            fLx = floor(Lxfit/split);

            % initialise output
            pfit = [];
            etafit = [];
            H = [];   

                for i = 0:(fLx-1),
                 % part of the fit
                    [pfit1,etafit1,H1] = feval(fun_estim,xfit(i*split + (1:split)),r,m,...
                         h(i*split + (1:split)),x,link,guessing,lapsing,K,p,ker,maxiter,tol);
                    % put the fits together
                    pfit = [pfit;pfit1];
                    etafit = [etafit;etafit1];
                    H = [H;H1];   
                end
                %final part of the fit
                if ((split*fLx)<Lxfit),
                    [pfit1,etafit1,H1] = feval(fun_estim,xfit((1+split*fLx):Lxfit),r,m,...
                        h((1+split*fLx):Lxfit),x,link,guessing,lapsing,K,p,ker,maxiter,tol);
                    % put the fits together
                    pfit = [pfit;pfit1];
                    etafit = [etafit;etafit1];
                    H = [H;H1];   
                end
        end
    else % no Hat matrix
        if (Lxfit<=split),

            % small x
            [pfit,etafit] = feval(fun_estim,xfit,r,m,h,x,link,guessing,lapsing,...
                K,p,ker,maxiter,tol);
        else
            % large x

            % number of parts into which the fitting is divided
            fLx = floor(Lxfit/split);

            % initialise output
            pfit = [];
            etafit = [];

            for i = 0:(fLx-1),
                % part of the fit
                [pfit1,etafit1] = feval(fun_estim,xfit(i*split + (1:split)),r,m,h(i*split + (1:split)),x,...
                    link,guessing,lapsing,K,p,ker,maxiter,tol);
                % put the fits together
                pfit = [pfit;pfit1];
                etafit = [etafit;etafit1];
            end
            %final part of the fit
            if ((split*fLx)<Lxfit),
                [pfit1,etafit1] = feval(fun_estim,xfit((1+split*fLx):Lxfit),r,m,h((1+split*fLx):Lxfit),x,...
                    link,guessing,lapsing,K,p,ker,maxiter,tol);

                % put the fits together
                pfit = [pfit;pfit1];
                etafit = [etafit;etafit1];
            end

        end
    
    end % if nargout == 3
end % if length( h ) == 1

File: pkg/R/locglmfit.R
locglmfit<-function( xfit, r, m, x, h, returnH = FALSE, link = c( "logit" ),
                     guessing = 0, lapsing = 0, K = 2, p = 1,
                     ker = c( "dnorm" ), maxiter = 50, tol = 1e-6 ) {
#
# Local polynomial estimator for the psychometric function and eta function (psychometric function 
# transformed by link) for binomial data; also returns the hat matrix H. Actual calculations are 
# done in LOCGLMFIT_PRIVATE or LOCGLMFIT_SPARSE_PRIVATE depending on the size of the data set. 
# Here, the data are split into several parts to speed up the calculations.
#
#
#INPUT
#
# xfit - points at which to calculate the estimate pfit
# r    - number of successes at points x
# m    - number of trials at points x 
# x    - stimulus levels 
# h    - bandwidth(s)
#
# OPTIONAL INPUT
#
# returnH  - logical, if TRUE then hat matrix is calculated; default is FALSE
# link     - name of the link function; default is 'logit'
# guessing - guessing rate; default is 0
# lapsing  - lapsing rate; default is 0
# K    - power parameter for Weibull and reverse Weibull link; default is 2
# p        - degree of the polynomial; default is 1
# ker      - kernel function for weights; default is 'dnorm'
# maxiter  - maximum number of iterations in Fisher scoring; default is 50
# tol      - tolerance level at which to stop Fisher scoring; default is 1e-6
#
# OUTPUT
#
# pfit    - value of the local polynomial estimate at points xfit
# etafit  - estimate of eta (link of pfit)
# H       - hat matrix (OPTIONAL)

# First 5 arguments are mandatory
    if( missing("xfit") || missing("r") || missing("m") || missing("x") || missing("h") ) {
        stop("Check input. First 5 arguments are mandatory");
    }

# CHECK ROBUSTNESS OF INPUT PARAMETERS
    if ( !is.vector( xfit ) ) {
        stop("Vector xfit should be a column vector");
    }
    checkdata<-list();
    checkdata[[1]] <- x;
    checkdata[[2]] <- r;
    checkdata[[3]] <- m;
    checkinput( "psychometricdata", checkdata );
    rm( checkdata )
    if ( !is.vector( h ) ) {
        stop("Bandwidths h should be a vector");
    }
    if( !( length( h ) == 1 || length( h ) == length( xfit ) ) ) {
        stop( "Bandwidth h must be either a scalar or a vector with the same number of elements as xfit" );
    }
    checkinput( "linkfunction", link );
    if( length( guessing ) > 1 ) {
        stop( "Guessing rate must be a scalar" );
    }
    if( length( lapsing ) > 1 ) {
        stop( "Lapsing rate must be a scalar" );
    }

    checkinput( 'guessingandlapsing', c(guessing, lapsing) );
   if (link == "weibull" || link == "revweibull"){
	    checkinput( "exponentk", K );
	    }
    pn <- list()
	pn[[1]] <- p
	pn[[2]] <- x
    checkinput( "degreepolynomial", pn );
    checkinput( 'kernel', ker );
    checkinput( 'maxiter', maxiter );
    checkinput( 'tolerance', tol );

# INITIAL VALUES
    split <- 20;

    Lxfit <- length(xfit);
    Lx <- length(x);

	 value <- NULL
    pfit <- NULL;
    etafit    <- NULL;
    if( returnH  ) H <- NULL;

	if( link == "logit"      ||
        link == "probit"     ||
        link == "loglog"     ||
        link == "comploglog" ||
        link == "weibull"    ||
        link == "revweibull" ) {
            	
    			link <- paste( link, "_link_private", sep = "" );
	}
	
    if( Lx > 15 ) {
# big data
# First try to load package SparseM
        options(warn = -1);
        existSparseM <- library( SparseM, logical.return = TRUE );
        options(warn = 0);
        if( existSparseM ) {
            fun_estim <- locglmfit_sparse_private;
        }
        else {
            fun_estim <- locglmfit_private;
            message("Package SparseM not installed. No sparse matrices used");
            message("SparseM can be found at CRAN web site http://cran.r-project.org/");
        }
    }
    else {
# small data
        fun_estim <- locglmfit_private;
    }

# SPLIT AND EVALUATION
################################################################ SCALAR h
    if( length( h ) == 1 ) {
# with Hat matrix
        if( returnH ) {
            if( Lxfit <= split ) {
# small x

                value <- fun_estim( xfit, r, m, x, h, returnH, link, guessing,
                       lapsing, K, p, ker, maxiter, tol);
            }
            else {
# large x
# number of parts into which the fitting is divided
                fLx = floor( Lxfit / split );
# initialise output
                    for( i in 0:(fLx-1) ) {
# part of the fit
                        value1 <- fun_estim( xfit[i*split+c(1:split)], r, m, x, h,
                                returnH, link, guessing, lapsing, K, p, ker,
                                maxiter, tol );
# put the fits together
                        pfit <- c( pfit, value1$pfit );
                        etafit    <- c( etafit,    value1$etafit );
                        H      <- rbind(H,   value1$H );
                    }
# final part of the fit
                    if( ( split * fLx ) < Lxfit ) {
                        value1 <- fun_estim( xfit[c(1+split*fLx):Lxfit], r, m, x, h,
                                returnH, link, guessing, lapsing, K, p, ker,
                                maxiter, tol );

# put the fits together
                        pfit <- c( pfit, value1$pfit );
                        etafit    <- c( etafit,    value1$etafit );
                        H      <- rbind(H,   value1$H );
                    }
#   values to return

value$pfit <- pfit
value$etafit <- etafit
value$H <- H

				}
        }
        else { # no Hat matrix

            if( Lxfit <= split ) {
# small x

                value <- fun_estim( xfit, r, m, x, h,returnH, link, guessing,
                       lapsing, K, p, ker, maxiter, tol );
            }
            else {
# large x
# number of parts into which the fitting is divided
                fLx = floor( Lxfit / split );

# initialise output
                for( i in 0:(fLx-1) ) {
# part of the fit

                    value1 <- fun_estim(  xfit[i*split + c(1:split)], r, m, x, h,
                                returnH, link, guessing, lapsing, K, p, ker,
                                maxiter, tol );
# put the fits together
                    pfit <- c( pfit, value1$pfit );
                    etafit    <- c( etafit,    value1$etafit );
                }
# final part of the fit
                if( ( split * fLx ) < Lxfit ) {
                    value1 <- fun_estim( xfit[c(1+split*fLx):Lxfit], r, m, x, h,
                            returnH, link, guessing, lapsing, K, p, ker,
                            maxiter, tol );
# put the fits together
                    pfit <- c( pfit, value1$pfit );
                    etafit    <- c( etafit,    value1$etafit );
                }
           
#   values to return

value$pfit <- pfit
value$etafit <- etafit

			}
        } # if( returnH )
    }
    
 ################################################################ VECTOR h   
    else { # if( length( h ) == 1 )
    	
# with Hat matrix
        if( returnH ) {
            if ( Lxfit <= split ) {
# small x
                value <- fun_estim( xfit, r, m, x, h, returnH, link, guessing,
                       lapsing, K, p, ker, maxiter, tol );
                  
            }
            else {
# large x
# number of parts into which the fitting is divided
                fLx = floor( Lxfit / split );
                    for( i in 0:(fLx-1) ) {
# part of the fit
                        value1 <- fun_estim( xfit[i*split + c(1:split)], r, m, x,
                                h[i*split + c(1:split)], returnH, link,
                                guessing, lapsing, K, p, ker, maxiter,
                                tol );
# put the fits together
                        pfit <- c( pfit, value1$pfit );
                        etafit    <- c( etafit,    value1$etafit );
                        H      <- rbind(H,   value1$H );
                    }
# final part of the fit
                    if( ( split * fLx ) < Lxfit ) {
                        value1 <- fun_estim( xfit[c(1+split*fLx):Lxfit], r, m, x,
                                h[c(1+split*fLx):Lxfit], returnH, link,
                                guessing, lapsing, K, p, ker, maxiter,
                                tol );
# put the fits together
                        pfit <- c( pfit, value1$pfit );
                        etafit    <- c( etafit,    value1$etafit );
                        H      <- rbind(H,   value1$H );
                    }
#   values to return

value$pfit <- pfit
value$etafit <- etafit
value$H <- H

				}        
        }# end if(retunH)
        else { # no Hat matrix
            if( Lxfit <= split ) {
# small x
                value <- fun_estim( xfit, r, m, x, h, returnH, link, guessing,
                       lapsing, K, p, ker, maxiter, tol );
            }
            else {
# large x
# number of parts into which the fitting is divided
            fLx = floor( Lxfit / split );

# initialise output
                for( i in 0:(fLx-1) ) {
# part of the fit
                    value1 <- fun_estim( xfit[i*split + c(1:split)], r, m, x,
                            h[i*split + c(1:split)], returnH, link,
                            guessing, lapsing, K, p, ker, maxiter, tol );
# put the fits together
                    pfit <- c( pfit, value1$pfit );
                    etafit    <- c( etafit,    value1$etafit );
                }
# final part of the fit
                if( ( split * fLx ) < Lxfit ) {
                    value1 <- fun_estim( xfit[c(1+split*fLx):Lxfit], r, m, x,
                            h[c(1+split*fLx):Lxfit], returnH, link,
                            guessing, lapsing, K, p, ker, maxiter, tol );
# put the fits together
                    pfit <- c( pfit, value1$pfit );
                    etafit    <- c( etafit,    value1$etafit );
                }

#   values to return

value$pfit <- pfit
value$etafit <- etafit

			}
        } # if( returnH )
        
    } # if( length( h ) == 1 )

   return( value )
}